### PR TITLE
Only apply grouped GEMM padding for MXFP8 and FP8 non-HybridEP cases

### PR DESCRIPTION
**Pull request by danielvegamyhre**
## Context
- BF16 grouped GEMM no longer requires padding, we can remove it from the BF16 path and only use it for FP8 and MXFP8 grouped GEMMs
- TorchTitan will now only only contain a torch native "rank major to expert major" permutation impl for BF16 grouped GEMM, and not any extra per group padding kernels/logic for FP8/MXFP8 (these will live in torchao, as the quantization library it is a better home for them).

## Summary
There are 7 cases to handle:

- Case 1: BF16 + NoEP  
    - (do nothing)
- Case 2: BF16 + EP
    - Torch native impl handles permute from rank major to expert major (no padding) 
- Case 3: MXFP8 + No EP
    - Handled with pad/unpad kernels in torchao   
- Case 4: MXFP8 + Standard EP
    - torchao `permute_and_pad()` if token_group_alignment_size > 0, in ExpertParallel implementation
- Case 5: MXFP8 + HybridEP
    - HybridEP handles token group padding for MXFP8 grouped GEMM as part of the all2all dispatch
- Case 6: FP8 + No EP
    - Same as case 3  
- Case 7: FP8 + EP
    - Same as case 4 

## Misc changes
- Delete `kernels.py`
- Delete tests for those kernels
- Remove `pad_token_groups_for_grouped_mm` option from MXFP8ConverterConfig, since we can set it correctly automatically 
- Added debug models for float8 and mxfp8 to config registry to speed up future development 

## Tests
- BF16 + No EP (ok): https://www.internalfb.com/phabricator/paste/view/P2244567028
- BF16 + EP (ok): https://www.internalfb.com/phabricator/paste/view/P2244568064
- MXFP8 + No EP (issue): https://www.internalfb.com/phabricator/paste/view/P2244573156
    - Important follow up: nan loss for this case! probably a bug in the new pad/unpad kernels we added for this in torchao, as they have unit tests but this was the first e2e training test 
- MXFP8 + EP (ok): https://www.internalfb.com/phabricator/paste/view/P2244569057
- FP8 + No EP (ok): https://www.internalfb.com/phabricator/paste/view/P2244583497
- FP8 + EP (ok): https://www.internalfb.com/phabricator/paste/view/P2244584583
- I did not test Case 5 (Hybrid EP + MXFP8) because it was recently tested by others, and is a different code path not part of these changes.

FP8 tests were done with fp8 grouped mm only, not fp8 linear. Using both I get this weird tyro error?

```
[rank0]:│ model-converters.converters.0:config was not a match because:                │
[rank0]:│ • Default value Config(enable_fsdp_float8_all_gather=False,                  │
[rank0]:│   precompute_float8_dynamic_scale_for_fsdp=False, recipe_name=None,          │
[rank0]:│   filter_fqns=['output', 'router.gate'], emulate=False) with type Config     │
[rank0]:│   does not match type <class 'torchtitan.components.quantization.float8.Floa │
[rank0]:│   t8GroupedMMConverter.Config'>         
```

### CODE REVIEW

**Comment by danielvegamyhre** [comment]
Have not tested yet because of devgpu issues but if you want to take a look feel free @tianyu-l

**Comment by danielvegamyhre** [comment]
I finished testing @tianyu-l this is ready for review

File: torchtitan/models/common/moe/utils.py [hunk 1]
```diff
@@ -12,37 +12,21 @@
 from torchtitan.components.quantization import MXFP8_GROUP_ALIGNMENT_SIZE
 from torchtitan.tools.utils import _round_up
 
-from .kernels import generate_permute_indices
+TOKEN_GROUP_ALIGN_SIZE_M = 0
```
Comment (tianyu-l) [inline torchtitan/models/common/moe/utils.py]: we should remove this -- setting global variables is error-prone

we should move logic to parallelize functions for various combinations

File: torchtitan/models/common/moe/utils.py [hunk 2]
```diff
@@ -55,38 +39,30 @@ def set_token_group_alignment_size_m(
...
-def _unpermute(out, input_shape, permuted_indices):
-    out_unpermuted = out.new_empty(input_shape)
-    out_unpermuted[permuted_indices, :] = out
-    out = out_unpermuted[:-1]
-    return out
+def maybe_align_num_tokens_for_mxfp8(num_tokens: int) -> int:
+    """Round up token count only when MXFP8 group alignment is active."""
+    if TOKEN_GROUP_ALIGN_SIZE_M != MXFP8_GROUP_ALIGNMENT_SIZE:
+        return num_tokens
+    return _round_up(num_tokens, MXFP8_GROUP_ALIGNMENT_SIZE)
```
Comment (tianyu-l) [inline torchtitan/models/common/moe/utils.py]: move this logic to hybridep.py, including `_round_up` (as an inline function) which is currently only used once in this repo

File: torchtitan/distributed/expert_parallel.py [hunk 3]
```diff
@@ -135,28 +140,54 @@ def _token_dispatch(
...
-        ) = _permute(
-            routed_input, num_tokens_per_expert_group, ep_degree, num_local_experts
-        )
-
-        return routed_input, num_tokens_per_expert_group
+        # Non-HybridEP case:
+        # FP8/MXFP8 require groups to be permuted to expert major order AND padded to
+        # `alignment_size`.
+        # Otherwise, we only need to permute to expert major order.
+        if self.token_group_alignment > 0:
```
Comment (tianyu-l) [inline torchtitan/distributed/expert_parallel.py]: IMO the proper way is to create e.g. `FP8ExpertParallel` and dispatch to it in parallelize function, instead of making if-else in existing ExpertParallel.

Also the condition should be whether quantization is used, not the `token_group_alignment` size set from somewhere.

File: torchtitan/models/common/moe/utils.py [hunk 4]
```diff
@@ -118,3 +94,67 @@ def wrapper(
...
+        )
+
+    input_shape = x.shape
+    x = x[permuted_indices, :]
+
+    return input_shape, x, permuted_indices, num_tokens_per_expert
+
+
+# Source: https://github.com/pytorch/torchtitan/pull/2255
+def _generate_permute_indices(
```
Comment (tianyu-l) [inline torchtitan/models/common/moe/utils.py:63]: could you verify that before vs. after, we get bitwise identical results under same seed and determinism?

File: torchtitan/models/common/moe/utils.py (hunk 1 above)
Comment (danielvegamyhre) [reply torchtitan/models/common/moe/utils.py]: makes sense, i am working on this, the changes are straightforward for standard EP, but for Hybrid EP it seems like it will require (1) refactoring the custom ops, DispatchState etc to pass around the quantization type used, or (2) just a module level variable storing the quantization type, similar to `_buffer`.  I think (2) is a less invasive change, wdyt?

File: torchtitan/distributed/expert_parallel.py (hunk 3 above)
Comment (danielvegamyhre) [reply torchtitan/distributed/expert_parallel.py]: That works, did a refactor

File: torchtitan/models/common/moe/utils.py [hunk 5]
```diff
@@ -5,116 +5,99 @@
...
 
-from torchtitan.components.quantization import MXFP8_GROUP_ALIGNMENT_SIZE
-from torchtitan.tools.utils import _round_up
-
-from .kernels import generate_permute_indices
-
-TOKEN_GROUP_ALIGN_SIZE_M = 8
-ValidTokenGroupAlignmentSize = Literal[8, 16, 32]
 
+def indices_padding_wrapper(func: Callable) -> Callable:
```
Comment (tianyu-l) [inline torchtitan/models/common/moe/utils.py:12]: I don't think we need this function any more. Please remove this and simplify https://github.com/pytorch/torchtitan/blob/c0c0bf965e32906b5e80f4f42b1f42ebe9851840/torchtitan/models/common/moe/moe.py#L120-L131

File: torchtitan/models/gpt_oss/moe.py [hunk 6]
```diff
@@ -45,10 +45,9 @@ def backward(ctx, grad_output):
 
 def indices_padding_wrapper(func: Callable) -> Callable:
```
Comment (tianyu-l) [inline torchtitan/models/gpt_oss/moe.py:46]: same

File: torchtitan/distributed/expert_parallel.py [hunk 7]
```diff
@@ -147,16 +151,251 @@ def _token_dispatch(
...
+            routed_input,
+            self.permuted_indices,
+            num_tokens_per_expert_group_padded,
+            group_offsets,
+        ) = permute_and_pad(
+            routed_input,
+            num_tokens_per_expert_group,
+            ep_degree,
+            num_local_experts,
+            FLOAT8_GROUP_ALIGNMENT_SIZE,
```
Comment (tianyu-l) [inline torchtitan/distributed/expert_parallel.py:267]: why do you need two different classes? You could just init with different quantization type, which can be used to determine the alignment size, e.g. based on a static dict.

File: torchtitan/components/quantization/__init__.py [hunk 8]
```diff
@@ -32,5 +32,5 @@ class Config(Configurable.Config):
 
 
 # Module level global constants
-FP8_GROUP_ALIGNMENT_SIZE = 16
+FLOAT8_GROUP_ALIGNMENT_SIZE = 16
 MXFP8_GROUP_ALIGNMENT_SIZE = 32
```
Comment (tianyu-l) [inline torchtitan/components/quantization/__init__.py:36]: make this a dict from quantization type to alignment size

File: torchtitan/components/quantization/utils.py [hunk 9]
```diff
@@ -4,7 +4,35 @@
...
+    return False
+
+
+def get_grouped_mm_quantization_type(
+    model_converters: list[ModelConverter],
+) -> QuantizationType | None:
+    if find_float8_grouped_mm_config(model_converters):
+        return QuantizationType.FLOAT8
+    elif config := find_mxfp8_config(model_converters):
+        if routed_experts_in_fqns(config.fqns):
```
Comment (tianyu-l) [inline torchtitan/components/quantization/utils.py:34]: no need to modularize into multiple small functions which are not used elsewhere -- we can make everything in a single util function for now

File: torchtitan/components/quantization/utils.py [hunk 10]
```diff
@@ -4,7 +4,35 @@
...
 
+from enum import auto, Enum
+
 import torch.nn as nn
+from torchtitan.components.quantization.float8 import find_float8_grouped_mm_config
+from torchtitan.components.quantization.mx import find_mxfp8_config
+from torchtitan.protocols import ModelConverter
+
+
+class QuantizationType(Enum):
```
Comment (tianyu-l) [inline torchtitan/components/quantization/utils.py:15]: you already have https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/quantization/mx.py#L35, can we just use the strings?

File: torchtitan/components/quantization/mx.py [hunk 11]
```diff
@@ -78,21 +70,38 @@ def __init__(
...
                 "torch.compile enablement is required for highest performance of MXFP8 dynamic quantization."
             )
 
+        logger.info("MXFP8 MoE training enabled")
+
+        # If EP is enabled, TorchTitan handles the token group padding for MXFP8 grouped GEMM
+        # as part of the EP implementation.
+        # Otherwise, if EP is not enabled, we need TorchAO to pad the token groups.
+        self.pad_token_groups_for_grouped_mm = not parallel_dims.ep_enabled
+        logger.warning(
```
Comment (tianyu-l) [inline torchtitan/components/quantization/mx.py:91]: why it's a warning? sounds like a comment to me, especially when both hybridEP is used this warning would still be there

File: torchtitan/distributed/deepep/hybridep.py [hunk 12]
```diff
@@ -404,6 +400,7 @@ def dispatch_tokens(
     group: ProcessGroup,
     score_before_experts: bool = True,
     non_blocking_expert_capacity_factor: float | None = None,
+    quantization_type: QuantizationType | None = None,
```
Comment (tianyu-l) [inline torchtitan/distributed/deepep/hybridep.py:403]: hybridep module doesn't need to know the quantization_type. All it needs to know is pad multiple size.

File: torchtitan/distributed/expert_parallel.py [hunk 13]
```diff
@@ -147,16 +151,251 @@ def _token_dispatch(
...
+            device_mesh,
+            partition_fn=self._partition_fn,
+            # pyrefly: ignore [bad-argument-type]
+            input_fn=self._token_dispatch,
+            # pyrefly: ignore [bad-argument-type]
+            output_fn=self._token_combine,
+        )
+
+
+class Float8ExpertParallel(BaseExpertParallel):
```
Comment (tianyu-l) [inline torchtitan/distributed/expert_parallel.py:184]: can you inherit `ExpertParallel` instead of `BaseExpertParallel`, which can save a lot of code?

**Comment by pianpwk** [comment]
Would it also be possible to remove the for-loop padding/unpadding path (for < SM90) in https://github.com/pytorch/torchtitan/blob/b01adfb544b4331ecab090ebdb50b2296cd8eb6a/torchtitan/models/common/moe/moe.py#L24 and https://github.com/pytorch/torchtitan/blob/b01adfb544b4331ecab090ebdb50b2296cd8eb6a/torchtitan/models/gpt_oss/moe.py#L99 , as part of this work?

I think that should resolve issues https://github.com/pytorch/torchtitan/issues/2312 and https://github.com/pytorch/torchtitan/pull/2741

**Comment by tianyu-l** [comment]
I can take this over if you are busy @danielvegamyhre

File: torchtitan/components/quantization/mx.py [hunk 14]
```diff
@@ -78,21 +70,38 @@ def __init__(
...
             logger.warning(
                 "torch.compile enablement is required for highest performance of MXFP8 dynamic quantization."
             )
 
+        logger.info("MXFP8 MoE training enabled")
+
+        # If EP is enabled, TorchTitan handles the token group padding for MXFP8 grouped GEMM
+        # as part of the EP implementation.
+        # Otherwise, if EP is not enabled, we need TorchAO to pad the token groups.
+        self.pad_token_groups_for_grouped_mm = not parallel_dims.ep_enabled
```
Comment (tianyu-l) [inline torchtitan/components/quantization/mx.py:90]: what happens for float8, where you don't have this flag?

**Comment by danielvegamyhre** [comment]
thank you @tianyu-l yes i have been busy with an urgent NaN loss situation reported by a customer... will take a look!

**Comment by danielvegamyhre** [comment]
Closing in favor of #2774, thanks @tianyu-l !
